Allow in0 and output to be sharded on different grids for mcast 2D matmuls

TT-BrianLiu opened the pull request:
### Ticket
- https://github.com/tenstorrent/tt-metal/issues/8536

### Problem description
- For matmul 2D, when in0 is sharded, we require the output grid to be on the same grid as the in0 shard grid. This greatly simplifies the mcasting logic, but restricts what users can do with `per_core_N` when we have sharded inputs. This PR removes that restriction, similar to what was done for the 1D case in: 392a717a9785acb71ad3ccfb953f0ad1d2de86de

### What's changed
- Fully decouple number of cores for in0 sender and receiver along rows (transpose_mcast=False) or columns (transpose_mcast=True)
- This means user can now width shard in0 K and specify per_core_N that divides output width on arbitrary number of cores
- See tests/ttnn/sweep_tests/sweeps/sweeps/matmul/short/matmul_user_program_config_mcast_2d.py for examples

Changes:
- Remove this assert: TT_FATAL(K / (shard_shape[1] / TILE_WIDTH) == div_up(N, program_config.per_core_N))
- Use new reader_bmm_tile_layout_in0_sender_receiver_padding_width_sharded.cpp reader for both 1D and 2D matmul
  - Rename reader back to reader_bmm_tile_layout_in0_sender_receiver_padding_block_sharded.cpp
- Similar changes to 1D matmul (commit: 392a717) except:
  - Only two quadrants for sharded in0 readers: in0_mcast_cores_with_work_and_in_receiver_grid and in0_mcast_cores_without_work_and_not_in_receiver_grid
  - Additional complexity to handle transpose_mcast=False and transpose_mcast=True
- Update tests/ttnn/sweep_tests/sweeps/sweeps/matmul/short/matmul_user_program_config_mcast_1d.py to avoid code duplicating when sweeping in0_block_w values

### Checklist
- [x] Post commit CI passes
- [x] Model regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

TT-BrianLiu committed:
#0: Add C++ matmul program config members to python class

bbradelTT commented on tt_eager/tt_dnn/op_library/bmm/multi_core_reuse_mcast_2d_optimized/bmm_op_multi_core_reuse_mcast_2d_optimized.cpp:
```diff
@@ -113,77 +112,145 @@ operation::ProgramWithCallbacks create_program_mcast_in0_in1(
+        // in0 shard grid already accounts for transpose_mcast
+        // ie. If transpose_mcast, in0 width is along y
+        in0_sender_num_cores_along_width = transpose_mcast ? in0_shard_grid.y : in0_shard_grid.x;
+        if (in0_sender_num_cores_along_width > num_blocks_x) {
```
Under what circumstances is `in0_sender_num_cores_along_width <= num_blocks_x` and what happens further in the code?

TT-BrianLiu replied:
It happens when in0 shard width is less than the number of cores along output width. The code already works for this case since we used to mcast to the full output grid anyways. But now, it's cleaner and more explicit

bbradelTT replied:
Ok. Thanks.

bbradelTT commented on tt_eager/tt_dnn/op_library/bmm/multi_core_reuse_mcast_2d_optimized/bmm_op_multi_core_reuse_mcast_2d_optimized.cpp:
```diff
@@ -113,77 +112,145 @@ operation::ProgramWithCallbacks create_program_mcast_in0_in1(
+    uint32_t in0_sender_num_cores_along_width;
+
+    // Only used for in0 block sharded
+    CoreRange in0_mcast_cores_with_work_and_in_receiver_grid(
```
Instead of these two placeholders, why not define these variables, with the right values set from the start, further down in the if block you added around line 462 `if (in0_block_sharded) {` which at first glance looks like the only code that uses them?

TT-BrianLiu replied:
We can definitely move code like this together inside the same namespace. Problem is, in our c++ op host code, we typically have chunks of code that does something. In this case, we have:
- set up some `CoreRangeSet`s to describe regions on our device
- set up bunch of compile time args for reader kernels
- create the kernels using `CoreRangeSet` and compile time args

I think this makes the code a bit more readable. OR, we can move everything together (so fully create each kernel for each case). That also makes sense to me, but would probably be better in a separate PR.

bbradelTT replied:
Ok. Leaving it as is in this PR is fine.

TT-BrianLiu committed:
#8536: Allow in0 and output to be sharded/produced on different grids for mcast 2D
- Fully decouple number of cores for in0 sender and receiver along rows (transpose_mcast=False) or columns (transpose_mcast=True)
- This means user can now width shard in0 K and specify per_core_N that divides output width on arbitrary number of cores
- See tests/ttnn/sweep_tests/sweeps/sweeps/matmul/short/matmul_user_program_config_mcast_2d.py for examples

Changes:
- Remove this assert: TT_FATAL(K / (shard_shape[1] / TILE_WIDTH) == div_up(N, program_config.per_core_N))
- Use new reader_bmm_tile_layout_in0_sender_receiver_padding_width_sharded.cpp reader for both 1D and 2D matmul
  * Rename reader back to reader_bmm_tile_layout_in0_sender_receiver_padding_block_sharded.cpp
- Similar changes to 1D matmul (commit: 392a717) except:
  * Only two quadrants for sharded in0 readers: in0_mcast_cores_with_work_and_in_receiver_grid and in0_mcast_cores_without_work_and_not_in_receiver_grid
  * Additional complexity to handle transpose_mcast=False and transpose_mcast=True
- Update tests/ttnn/sweep_tests/sweeps/sweeps/matmul/short/matmul_user_program_config_mcast_1d.py to avoid code duplicating when sweeping in0_block_w values

TT-BrianLiu commented:
Passing pipeline:
- post-commit all: https://github.com/tenstorrent/tt-metal/actions/runs/9486245468
- post-commit ttnn unit tests: https://github.com/tenstorrent/tt-metal/actions/runs/9486247039
- post-commit model tests: https://github.com/tenstorrent/tt-metal/actions/runs/9486250335
- nightly fast dispatch: https://github.com/tenstorrent/tt-metal/actions/runs/9486252928 (failures are same as on `main`)

tt-aho reviewed (approved):
lgtm
